core/trino-main/src/main/java/io/trino/execution/executor2/scheduler/SchedulingQueue.java

```diff
@@ -0,0 +1,531 @@
+                case RUNNING, RUNNABLE -> builder.append("  %s %s [%s, weight = %s, baseline = %s]\n".formatted(
+                        group == runnableQueue.peek() ? "=>" : " -",
+                        group,
+                        info.state(),
+                        info.weight(),
+                        info.baselineWeight()));
+            }
+
+            for (Map.Entry<T, Task> taskEntry : info.tasks.entrySet()) {
+                T task = taskEntry.getKey();
+                Task taskInfo = taskEntry.getValue();
+
+                if (info.blocked.containsKey(task)) {
+                    builder.append("         %s [BLOCKED, saved delta = %s]\n".formatted(task, info.blocked.get(task).savedDelta()));
+                }
+                else if (info.runnableQueue.contains(task)) {
+                    builder.append("      %s %s [RUNNABLE, weight = %s]\n".formatted(
+                            task == info.peek() ? "=>" : "  ",
+                            task,
+                            taskInfo.weight()));
+                }
+                else {
+                    builder.append("         %s [RUNNING, weight = %s, uncommitted = %s]\n".formatted(
+                            task,
+                            taskInfo.weight(),
+                            taskInfo.uncommittedWeight()));
+                }
+            }
+        }
+
+        return builder.toString();
+    }
+
+    private static class Group<T>
+    {
+        private State state;
+        private long weight;
+        private final Map<T, Task> tasks = new HashMap<>();
+        private final PriorityQueue<T> runnableQueue = new PriorityQueue<>();
+        private final Map<T, Blocked> blocked = new HashMap<>();
+        private final PriorityQueue<T> baselineWeights = new PriorityQueue<>();
+
+        public Group(long weight)
+        {
+            this.state = BLOCKED;
+            this.weight = weight;
+        }
+
+        public void enqueue(T task, long deltaWeight)
+        {
+            Task info = tasks.get(task);
+
+            if (info == null) {
+                // New tasks get assigned the baseline weight so that they don't monopolize the queue
+                // while they catch up
+                info = new Task(baselineWeight());
+                tasks.put(task, info);
+            }
+            else if (blocked.containsKey(task)) {
+                Blocked blockedTask = blocked.remove(task);
+                info.adjustWeight(blockedTask.savedDelta());
+            }
+
+            weight -= info.uncommittedWeight();
+            weight += deltaWeight;
+
+            info.commitWeight(deltaWeight);
+            runnableQueue.add(task, info.weight());
+            baselineWeights.addOrReplace(task, info.weight());
+
+            updateState();
+        }
+
+        public T dequeue(long expectedWeight)
+        {
+            checkArgument(state == RUNNABLE);
+
+            T task = runnableQueue.poll();
```

Comment:
Consider adding a method that is like `takeOrThrow` so it is clear to the reader this cannot return null